### ISSUE TITLE
could not convert to integer . Path ' exitCode'. value too big or too small for Int32

### ISSUE THREAD

**Issue by sgtkellox** [libvips/pyvips#428]
Hello,
I use your nice library to stitch some jpg tiles to a big tiff.
On most cases, it works fine. However sometimes the tiff generated is of size 0 bytes. If i run my code from command line, no error is produced. If I run it from visual studio, I get the following Error: 
```
could not convert to integer . Path ' exitCode'. value too big or too small for Int32
```
Here is the core of the  program I wrote: 
```
import os
import numpy as np


add_dll_dir = getattr(os, "add_dll_directory", None)
vipsbin = r"C:\AI\vips-dev-8.14\bin"
if callable(add_dll_dir):
    add_dll_dir(vipsbin)    
else:
    os.environ["PATH"] = os.pathsep.join((vipsbin, os.environ["PATH"]))
   
format_to_dtype = {
    'uchar': np.uint8,
    'char': np.int8,
    'ushort': np.uint16,
    'short': np.int16,
    'uint': np.uint32,
    'int': np.int32,
    'float': np.float32,
    'double': np.float64,
    'complex': np.complex64,
    'dpcomplex': np.complex128,
}

# map np dtypes to vips
dtype_to_format = {
    'uint8': 'uchar',
    'int8': 'char',
    'uint16': 'ushort',
    'int16': 'short',
    'uint32': 'uint',
    'int32': 'int',
    'float32': 'float',
    'float64': 'double',
    'complex64': 'complex',
    'complex128': 'dpcomplex',
}


import pyvips


#list the tiles to be stitched 
imgs = os.listdir(pathToMyTiles)


#slide dimensions are extracted from the filenames of the tiles
slideWidth,slideHeight = getWidthAndHeight(imgs)


result = pyvips.Image.black(slideWidth,slideHeight,bands=3)
print("result format " + result.format)


for img in imgs: 

    print("-----------------")

    tile = pyvips.Image.new_from_file(img, access='sequential')
    mem_img = tile.write_to_memory()

    print("tile format " + str(tile.format))
    print("tile dtype " + str(format_to_dtype[tile.format]))

    #convert to numpy
    np_3d = np.ndarray(buffer=mem_img,dtype=format_to_dtype[tile.format],shape=[tile.height, tile.width, tile.bands])
    height, width, bands = np_3d.shape
    linear = np_3d.reshape(width * height * bands)


    print("Numpy array DataType " + str(np_3d.dtype))
    print("numpy array format format "+ str(dtype_to_format[str(np_3d.dtype)]))

    #make some changes
    np_3d[:,:,0] = 220

    #reconvert to vips image
    vi = pyvips.Image.new_from_memory(linear.data, width, height, bands,
                                          dtype_to_format[str(np_3d.dtype)])

    print("format after going back to vips  " + str(vi.format))
    

    

    #tile position is extracted from the filename 
    absX,absY = extractTileCoordinates(img)

    result = result.insert(vi,absX,absY)

    result.tiffsave(safePath, compression=pyvips.enums.ForeignTiffCompression.DEFLATE,
                 tile=True, tile_width=512, tile_height=512, #rgbjpeg=True,
                 pyramid=True,  bigtiff=True)

    print("-----------------")
```
Since i suspected this to be some sort of datatype mismatch between my tiles, the numpy part or the final tiff, I printed those datatypes for inspection. Here is the output of a failed example. It is the same for all tiles in the tile folder. It seems all fine to me. 

```
result format uchar
-----------------
tile format uchar
tile dtype <class 'numpy.uint8'>
Numpy array DataType uint8
numpy array format format uchar
format after going back to vips  uchar
-----------------
```
Also maybe I should note, that my tiffs are really big (190k x  100k).
If possible, I would appreciate some advide on this problem.
Thanks in advance,
Felix

**Comment by jcupitt** [comment]
Hi @sgtkellox,

You don't need any of the numpy conversion stuff now, you can just use:

```python
numpy_array = vips_image.numpy()
```

To go from vips to numpy, and:

```python
vips_image = pyvips.Image.new_from_array(numpy_array)
```

to go back the other way. They are both efficient, ie. there is no copying of large memory buffers.

Instead of:

```python
    #make some changes
    np_3d[:,:,0] = 220
```

You can just do this in pyvips and save going to numpy and back, it'll be a lot quicker:

```python
# make a one-band constant image with each pixel having the value "220"
i220 = image.new_from_image(220)
# use that for red, but take g and b from image
image = i220.bandjoin([image[1:3])
```

So I would write:

```python
result = pyvips.Image.black(slideWidth, slideHeight, bands=3)

for img in imgs: 
    print("-----------------")
    tile = pyvips.Image.new_from_file(img, access='sequential')

    # set red band to 220
    tile = tile.new_from_image(220).bandjoin(tile[1:3])

    # tile position is extracted from the filename 
    absX,absY = extractTileCoordinates(img)

    result = result.insert(tile, absX, absY)

result.tiffsave(safePath, compression='deflate', 
                  tile=True, tile_width=512, tile_height=512, 
                  pyramid=True,  bigtiff=True)
```

And it ought to work for very large images, though it does depend on the exact format of your TIFF source images.

**Comment by jcupitt** [comment]
TIFF tilesizes need to be a multiple of 16.

Yes, I would use JPEG compression. Deflate is as slow as molasses, files are HUGE, and it's not supported by many programs. JPEG2000 is not very useful (imo).

**Comment by jcupitt** [comment]
I made you a tiny demo prog:

```python
#!/usr/bin/env python

import sys
import random
import pyvips

slide = pyvips.Image.black(10000, 10000, bands=3)

for filename in sys.argv[2:]:
    tile = pyvips.Image.new_from_file(filename, access="sequential")
    x = random.randint(0, slide.width - tile.width)
    y = random.randint(0, slide.height - tile.height)
    slide = slide.insert(tile, x, y)
    
slide.write_to_file(sys.argv[1])
```

It just inserts at random positions, and it's only 10k x 10k pixels, but it'll scale up very easily.

I can run it like this:

```
$ for i in {1..10}; do cp ~/pics/k2.jpg $i.jpg; done
$ ../sgtkellox.py x.tif[compression=jpeg,tile,tile-width=512,tile-height=512,pyramid] *.jpg
```

And it'll make a pyramid that loads into qupath 0.4.4 like this:

```
$ ./QuPath ~/try/sample/x.tif
```

![image](https://github.com/libvips/pyvips/assets/580843/49410bc9-4abd-4582-a9a9-fbc4a0bde469)

You can see it's been loaded via openslide, and all the pyramid levels have been detected correctly.

**Comment by jcupitt** [comment]
Just for fun, I tried changing it to:

```python
slide = pyvips.Image.black(1000000, 1000000, bands=3)
```

ie. a 1m x 1m pixel image. It ran like this:

```
$ /usr/bin/time -f %M:%e ../sgtkellox.py x.tif[compression=jpeg,tile,tile-width=512,tile-height=512,pyramid,bigtiff] *.jpg
7033768:18052.62
$ ls -l x.tif
-rw-r--r-- 1 john john 21120246588 Oct 11 23:43 x.tif
```

So that's a peak of 7gb of memory, it ran for 18,000s (about 5h), and wrote a 21gb file. It loaded into qupath OK:

![image](https://github.com/libvips/pyvips/assets/580843/9cd763b5-f9f0-4ac3-bd4e-de692c33919c)

**Comment by sgtkellox** [comment]
Thank you. 
So after doing a reinstall I got rid of the error when using JPEG compression.  I ran your script and mine on tiles of 25 different wsis of very different sizes. As you said JPEG is best, deflate makes to big files and is to slow, and JP2K produces wrong colored output. 
My stuff works now as intended. 
The thing that may be of interest for others is that you get the Error "
could not convert to integer . Path ' exitCode'. value too big or too small for Int32" if you try to insert tiles at positions exceeding the dimensions of the result tiff. As said above you get the error while saving, not while inserting. 
Thanks again, ill close the Issue now. 
Regards,
Felix
